src/dynsight/_internal/trajectory/trajectory.py

```diff
@@ -0,0 +1,202 @@
+from __future__ import annotations
+
+import pickle
+from dataclasses import dataclass
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import MDAnalysis
+    import numpy as np
+    from numpy.typing import NDArray
+    from tropea_clustering._internal.first_classes import StateMulti, StateUni
+
+import dynsight
+
+UNIVAR_DIM = 2
+
+
+@dataclass
+class Insight:
+    """Contains an analysis perfomed on a many-body trajectory.
+
+    Attributes:
+        dataset: np.ndarray of shape (n_particles, n_frames, n_dims)
+            The values of a single-particle descriptor.
+        r_cut: float
+            The spatial cutoff with the descriptor has been computed.
+    """
+
+    dataset: NDArray[np.float64]
+    r_cut: float
+
+    def spatial_average(
+        self,
+        trajectory: Trj,
+        r_cut: float,
+        selection: str = "all",
+        num_processes: int = 1,
+    ) -> Insight:
+        averaged_dataset = dynsight.analysis.spatialaverage(
+            universe=trajectory.universe,
+            descriptor_array=self.dataset,
+            selection=selection,
+            cutoff=r_cut,
+            num_processes=num_processes,
+        )
+        return Insight(averaged_dataset, self.r_cut)
+
+    def get_onion(self, delta_t: int) -> OnionInsight:
+        if self.dataset.ndim == UNIVAR_DIM:
+            reshaped_data = dynsight.onion.helpers.reshape_from_nt(
+                self.dataset, delta_t
+            )
+            onion_clust = dynsight.onion.OnionUni()
+        else:
+            reshaped_data = dynsight.onion.helpers.reshape_from_dnt(
+                self.dataset.transpose(2, 0, 1), delta_t
+            )
+            onion_clust = dynsight.onion.OnionMulti(ndims=self.dataset.ndim)
+
+        onion_clust.fit(reshaped_data)
+
+        return OnionInsight(
+            dataset=self.dataset,
+            r_cut=self.r_cut,
+            labels=onion_clust.labels_,
+            delta_t=delta_t,
+            state_list=onion_clust.state_list_,
+            reshaped_data=reshaped_data,
+        )
+
+    def save(self, file_name: str) -> None:
+        file_path = Path(f"{file_name}.pkl")
+        with Path.open(file_path, "wb") as file:
+            pickle.dump(self, file)
+
+
+@dataclass
+class ClusterInsight(Insight):
```

Comment:
I do not think we need to inherit Insight - because doing so means that they have to have the same attributes and such, right? like `r_cut`?